src/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs

```diff
@@ -135,7 +135,10 @@ private IEnumerator<T> GetEnumerator()
                 EnumerableRewriter rewriter = new EnumerableRewriter();
                 Expression body = rewriter.Visit(_expression);
                 Expression<Func<IEnumerable<T>>> f = Expression.Lambda<Func<IEnumerable<T>>>(body, (IEnumerable<ParameterExpression>)null);
-                _enumerable = f.Compile()();
+                IEnumerable<T> enumerable = f.Compile()();
+                if (enumerable == this)
+                    throw Error.EnumeratingNullEnumerableExpression();
```

Comment:
When the query is created would be ideal, but because the same class serves as both query and query provider there could be working code that creates a query from a null enumerable and then only ever uses it as a provider, and throwing on creation would hence break running code.
The check could be replaced with a null check in this method, but it wouldn't make much difference in the common case, and this way if there's some other pattern that could have the same effect it would also be caught.